File: icpc/24q1/mcna-2007/here-we-go(relians)-again (unsolved)/solution.cs
using System;
using System.Collections.Generic;
using System.Linq;

struct Link {
    public Tuple<int, int> destination;
    public int timeBlips;
}

class Program {
    private static int findBestRouteTimeBlips(
        Dictionary<Tuple<int, int>, List<Link>> links,
        Tuple<int, int> origin,
        Tuple<int, int> destination
    ) {
        var costs = new Dictionary<Tuple<int, int>, int>();
        var front = new HashSet<Tuple<int, int>>(); 
        
        costs.Add(origin, 0);
        front.Add(origin);

        while (front.Count > 0) {
            var nextFront = new HashSet<Tuple<int, int>>();

            foreach (var intersection in front) {
                foreach (var link in links.GetValueOrDefault(intersection, new List<Link>())) {
                    var oldCost = costs.GetValueOrDefault(link.destination, int.MaxValue);
                    var newCost = costs[intersection] + link.timeBlips;

                    if (newCost < oldCost) {
                        costs[link.destination] = newCost;
                        
                        nextFront.Add(link.destination);
                    }
                }
            }

            front = nextFront;
        }

        return costs.GetValueOrDefault(destination, -1);
    }

    public static void Main() {
        while (true) {
            var dimensions = readDimensions();

            if (dimensions.Equals(Tuple.Create(0, 0))) return;

            var rows = dimensions.Item2;

            var links = readRoads(rows);

            var origin = Tuple.Create(0, 0);
            var destination = dimensions;

            var bestRouteTimeBlips =
                findBestRouteTimeBlips(links, origin, destination);
 
            if (bestRouteTimeBlips == -1)
                Console.WriteLine("Holiday");

            if (bestRouteTimeBlips != -1)  
                Console.WriteLine($"{bestRouteTimeBlips} blips");
        }
    }

    private static readonly int STREET_LENGTH_RELS = 2520;

    private static readonly HashSet<char> H_DIRECTIONS =
        new HashSet<char>{'<', '>'};

    private static readonly HashSet<char> V_DIRECTIONS =
        new HashSet<char>{'^', 'v'};

    private static Tuple<int, int> readDimensions() {
        var dimensions = Console.ReadLine().Split(' ').Select(v => Convert.ToInt32(v));
        return Tuple.Create(dimensions.First(), dimensions.Last());
    }

    private static Dictionary<Tuple<int, int>, List<Link>> readRoads(int rows) {        
        var links = new Dictionary<Tuple<int, int>, List<Link>>();

        for (var i = 0; i < rows * 2 + 1; i++) {
            var y = i / 2;

            var values = Console.ReadLine().Split(' ');
            
            var readingHorizontalStreets = i % 2 == 0;
            var readingVerticalStreets = !readingHorizontalStreets;

            for (var j = 0; j < values.Length; j += 2) {
                var x = j / 2;
 
                var speed_rels_per_blip = Convert.ToInt32(values[j]);

                if (speed_rels_per_blip == 0) continue;

                var streetDirection = Convert.ToChar(values[j + 1]);

                IEnumerable<char> directions = default(HashSet<char>);

                if (readingHorizontalStreets) directions = H_DIRECTIONS;
                if (readingVerticalStreets) directions = V_DIRECTIONS;

                foreach (var direction in directions) {
                    if (streetDirection != direction && streetDirection != '*')
                        continue;

                    Tuple<int, int> origin = null;
                    Tuple<int, int> destination = null;

                    if (direction == '^') {
                        origin = Tuple.Create(x, y + 1);
                        destination = Tuple.Create(x, y);
                    }

                    if (direction == 'v') {
                        origin = Tuple.Create(x, y);
                        destination = Tuple.Create(x, y + 1);
                    }

                    if (direction == '<') {
                        origin = Tuple.Create(x + 1, y);
                        destination = Tuple.Create(x, y);
                    }

                    if (direction == '>') {
                        origin = Tuple.Create(x, y);
                        destination = Tuple.Create(x + 1, y);
                    }

                    if (origin == null || destination == null)
                        throw new Exception("Should never happen");

                    if (!links.ContainsKey(origin))
                        links[origin] = new List<Link>();

                    Link link;

                    link.destination = destination;
                    link.timeBlips = STREET_LENGTH_RELS / speed_rels_per_blip;

                    links[origin].Add(link);
                }
            }
        }

        return links;
    }
}

File: icpc/24q1/mcna-2007/here-we-go(relians)-again (unsolved)/solution.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <limits.h>
#include <stdint.h>

struct Link
{
    int x;
    int y;
    int timeBlips;
};

void readDimensions(int*, int*);

void readStreets(
    int columns,
    int rows,
    int linkCounts[columns + 1][rows + 1],
    struct Link* links[columns + 1][rows + 1]
);

int STREET_LENGTH_RELS = 2520;
char H_DIRECTIONS[] = {'<', '>'};
char V_DIRECTIONS[] = {'^', 'v'};

int64_t findBestRouteTimeBlips(
    int rows,
    int columns,
    int linkCounts[columns + 1][rows + 1],
    struct Link* links[columns + 1][rows + 1]
) {
    int64_t costs[columns + 1][rows + 1];

    for (int x = 0; x < columns + 1; x++)
        for (int y = 0; y < rows + 1; y++)
            costs[x][y] = INT64_MAX;

    int frontMaxSize = (columns + 1) * (rows + 1);
    int* frontX = malloc(frontMaxSize * sizeof(int));
    int* frontY = malloc(frontMaxSize * sizeof(int));
    int* nextFrontX = malloc(frontMaxSize * sizeof(int));
    int* nextFrontY = malloc(frontMaxSize * sizeof(int));

    int originX = 0;
    int originY = 0;
    int destinationX = columns;
    int destinationY = rows;

    costs[originX][originY] = 0;
    frontX[0] = originX;
    frontY[0] = originY;

    int frontSize = 1;
    int nextFrontSize;

    while (frontSize > 0)
    {
        nextFrontSize = 0;

        for (int i = 0; i < frontSize; i++)
        {
            int x = frontX[i];
            int y = frontY[i];

            for (int j = 0; j < linkCounts[x][y]; j++)
            {
                int linkX = links[x][y][j].x;
                int linkY = links[x][y][j].y;
                int64_t linkTimeBlips = links[x][y][j].timeBlips;

                int64_t oldCost = costs[linkX][linkY];
                int64_t newCost = costs[x][y] + linkTimeBlips;

                if (newCost < oldCost) {
                    costs[linkX][linkY] = newCost;
                    nextFrontX[nextFrontSize] = linkX;
                    nextFrontY[nextFrontSize] = linkY;
                    nextFrontSize++;
                }
            }
        }

        frontSize = nextFrontSize;

        int* tmpX = frontX;
        int* tmpY = frontY;

        frontX = nextFrontX;
        frontY = nextFrontY;

        nextFrontX = tmpX;
        nextFrontY = tmpY;
    }

    return costs[destinationX][destinationY];
}

int main()
{
    while (true) {
        int columns, rows;

        readDimensions(&columns, &rows);

        if (columns == 0 && rows == 0) return 0;

        int linkCounts[columns + 1][rows + 1];
        struct Link* links[columns + 1][rows + 1];

        readStreets(columns, rows, linkCounts, links);

        int64_t blips = findBestRouteTimeBlips(rows, columns, linkCounts, links);

        if (blips == INT64_MAX) printf("Holiday\n");
        if (blips != INT64_MAX) printf("%lld blips\n", blips);
    }

    return 0;
}

void readDimensions(int* x_ptr, int* y_ptr)
{
    scanf("%d %d", x_ptr, y_ptr);
}

void readStreets(
    int columns,
    int rows,
    int linkCounts[columns + 1][rows + 1],
    struct Link* links[columns + 1][rows + 1]
) {
    for (int y = 0; y < rows + 1; y++)
        for (int x = 0; x < columns + 1; x++)
            linkCounts[x][y] = 0;

    for (int i = 0; i < rows * 2 + 1; i++)
    {
        int y = i / 2;

        bool readingHorizontalStreets = i % 2 == 0;
        int streetsToRead = readingHorizontalStreets ? columns : columns + 1;

        for (int x = 0; x < streetsToRead; x++)
        {
            int streetSpeedRelsPerBlip;
            char streetDirection;

            scanf("%d %c", &streetSpeedRelsPerBlip, &streetDirection);

            if (streetSpeedRelsPerBlip == 0) continue;

            char* directions =
                readingHorizontalStreets ? H_DIRECTIONS : V_DIRECTIONS;

            for (int j = 0; j < 2; j++)
            {
                char direction = directions[j];

                if (streetDirection != direction && streetDirection != '*')
                    continue;

                int originX;
                int originY;
                int destinationX;
                int destinationY;

                if (direction == '^')
                {
                    originX = x;
                    originY = y + 1;
                    destinationX = x;
                    destinationY = y;
                }

                if (direction == 'v')
                {
                    originX = x;
                    originY = y;
                    destinationX = x;
                    destinationY = y + 1;
                }

                if (direction == '<')
                {
                    originX = x + 1;
                    originY = y;
                    destinationX = x;
                    destinationY = y;
                }

                if (direction == '>')
                {
                    originX = x;
                    originY = y;
                    destinationX = x + 1;
                    destinationY = y;
                }

                int originLinksCount = linkCounts[originX][originY];

                if (originLinksCount == 0)
                    links[originX][originY] = malloc(4 * sizeof(struct Link));

                struct Link* link = &links[originX][originY][originLinksCount];

                link->x = destinationX;
                link->y = destinationY;
                link->timeBlips = STREET_LENGTH_RELS / streetSpeedRelsPerBlip;

                linkCounts[originX][originY]++;
            }
        }
    }
}